Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -804,6 +804,113 @@ static ErrorOr<void> decode_jpeg2000_header(JPEG2000LoadingContext& context, Rea
     return {};
 }
 
+namespace JPEG2000 {
+
+// Tag trees are used to store the code-block inclusion bits and the zero bit-plane information.
+// B.10.2 Tag trees
+// "At every node of this tree the minimum integer of the (up to four) nodes below it is recorded. [...]
+//  Level 0 is the lowest level of the tag tree; it contains the top node. [...]
+//  Each node has a [...] current value, [...] initialized to zero. A 0 bit in the tag tree means that the minimum
+//  (or the value in the case of the highest level) is larger than the current value and a 1 bit means that the minimum
+//  (or the value in the case of the highest level) is equal to the current value.
+//  For each contiguous 0 bit in the tag tree the current value is incremented by one.
+//  Nodes at higher levels cannot be coded until lower level node values are fixed (i.e, a 1 bit is coded). [...]
+//  Only the information needed for the current code-block is stored at the current point in the packet header."
+// The example in Figure B.13 / Table B.5 is useful to understand what exactly "only the information needed" means.
+struct TagTreeNode {
+    u32 value { 0 };
+    enum State {
+        Pending,
+        Final,
+    };
+    State state { Pending };
+    Array<TagTreeNode*, 4> children {};
+    u32 level { 0 }; // 0 for leaf nodes, 1 for the next level, etc.
+
+    bool is_leaf() const { return level == 0; }
+
+    ErrorOr<u32> read_value(u32 x, u32 y, Function<ErrorOr<bool>()> read_bit, u32 start_value, Optional<u32> stop_at = {})
```

Comment:
It is hard to judge as I don't have the context and this is never used in practice, but that API doesn't feel right as taking a `Function` seems to only complicate things. 
Already in the test case, you need a factory to recreate the bit reader every single time. As far as I can see, you will need to do something similar if you need to read some values in a loop (which I expect to happen).
I think that taking a `InputBitStream` would be better here. (That `Function` is probably atrocious for performance as well, but I know that's not a concern at first.)

Reply:
I changed it to take the function by const ref instead. I think that removes most of the awkwardness.

At least at the moment, this is used once per codeblock, which is kind of like a 32x32 block (except the size is configurable by the encoder, and it's in DWT space), so perf isn't critical yet. (Maybe we'll need this per sample later though.) For now, having a function makes writing tests nicer.

(In real life, "B.10.1 Bit-stuffing routine" describes that one bit after a `0xff` byte has to be skipped, so I have a wrapper lambda around the input bit stream to handle that. I could make a whole stream wrapper class for that, but a function seems a bit simpler to me.)